src/ripple/app/tx/impl/SetCron.cpp

```diff
@@ -0,0 +1,274 @@
+        {
+            JLOG(j_.warn()) << "SetCron: Cron object didn't exist.";
+            return tefBAD_LEDGER;
+        }
+
+        if (safe_cast<LedgerEntryType>(
+                sleCron->getFieldU16(sfLedgerEntryType)) != ltCRON)
+        {
+            JLOG(j_.warn()) << "SetCron: sfCron pointed to non-cron object!!";
+            return tefBAD_LEDGER;
+        }
+
+        if (!view.dirRemove(
+                keylet::ownerDir(id), (*sleCron)[sfOwnerNode], klOld, false))
+        {
+            JLOG(j_.warn()) << "SetCron: Ownerdir bad. " << id;
+            return tefBAD_LEDGER;
+        }
+
+        view.erase(sleCron);
+        adjustOwnerCount(view, sle, -1, j_);
+        sle->makeFieldAbsent(sfCron);
+    }
+
+    // if the operation is a delete (no delay or recur specified then stop
+    // here.)
+    if (isDelete)
+    {
+        view.update(sle);
+        return tesSUCCESS;
+    }
+
+    // execution to here means we're creating a new Cron object and adding it to
+    // the user's owner dir
+
+    Keylet klCron = keylet::cron(afterTime, id);
+
+    bool const alreadyExists = view.exists(klCron);
+
+    std::shared_ptr<SLE> sleCron =
+        alreadyExists ? view.peek(klCron) : std::make_shared<SLE>(klCron);
+
+    if (!alreadyExists)
+    {
+        STAmount const reserve{
+            view.fees().accountReserve(sle->getFieldU32(sfOwnerCount) + 1)};
+
+        STAmount const afterFee =
+            mPriorBalance - ctx_.tx.getFieldAmount(sfFee).xrp();
+
+        if (afterFee > mPriorBalance || afterFee < reserve)
+            return tecINSUFFICIENT_RESERVE;
+
+        // add to owner dir
+        auto const page =
+            view.dirInsert(keylet::ownerDir(id), klCron, describeOwnerDir(id));
+        if (!page)
+            return tecDIR_FULL;
+
+        sleCron->setFieldU64(sfOwnerNode, *page);
+
+        adjustOwnerCount(view, sle, 1, j_);
+    }
+
+    // set the fields
+    sleCron->setFieldU32(sfDelaySeconds, delay);
+    sleCron->setFieldU32(sfRepeatCount, recur);
+    sleCron->setAccountID(sfOwner, id);
+
+    sle->setFieldH256(sfCron, klCron.key);
+
+    view.update(sle);
+
+    if (alreadyExists)
+        view.update(sleCron);
+    else
+        view.insert(sleCron);
+
+    return tesSUCCESS;
+}
+
+XRPAmount
+SetCron::calculateBaseFee(ReadView const& view, STTx const& tx)
+{
+    auto const baseFee = Transactor::calculateBaseFee(view, tx);
+
+    auto const hasRepeat = tx.isFieldPresent(sfRepeatCount);
+    auto const hasDelay = tx.isFieldPresent(sfDelaySeconds);
+
+    if (!hasRepeat && !hasDelay)
+        // delete cron
+        return baseFee;
+
+    // factor a cost based on the total number of txns expected
+    // for RepeatCount of 0 we have this txn (SetCron) and the
+    // single Cron txn (2). For a RepeatCount of 1 we have this txn,
+    // the first time the cron executes, and the second time (3).
+    uint32_t const additionalExpectedExecutions =
+        tx.getFieldU32(sfRepeatCount) + 1;
```

Comment:
this is a crash bug @tequdev sfRepeatCount may be absent indicating implicit 0 repeats